compliance_checker/cf/cf.py

```diff
@@ -2360,67 +2360,68 @@ def check_packed_data(self, ds):
+                valid_missing_variable = False
+
+                if hasattr(var.add_offset, 'dtype') and hasattr(var.scale_factor, 'dtype'):
+
+                    if var.dtype == var.add_offset.dtype == var.scale_factor.dtype:
                         valid = True
-                        result = Result(BaseCheck.MEDIUM,                            \
-                                valid,                                       \
-                                ('var', name, 'packed_data'), \
-                                reasoning)
-                        ret_val.append(result)
-                        reasoning = []
-                    elif type(scale) == type(offset) != type(np.reshape(ds.dataset.variables[name],data_size)[x]):
-                        if type(scale) in [float(), type(np.float32(1.)), int(), type(np.int16(1))]  and (type(np.reshape(ds.dataset.variables[name],data_size)[x])) in [type(np.int8(1)), type(np.int16(1)), int()]:
-                           valid = True
-                           result = Result(BaseCheck.MEDIUM,                            \
-                                valid,                                       \
-                                ('var', name, 'packed_data'), \
-                                reasoning)
-                           ret_val.append(result)
-                           reasoning = []
-                        else: 
-                           valid = False
-
-                           reasoning.append("'add_offset' and 'scale_factor' are not of type float or int, or the data variable is not of type byte, short, or int.")
-                           result = Result(BaseCheck.MEDIUM,                            \
-                                valid,                                       \
-                                ('var', name, 'packed_data'), \
-                                reasoning)
-                           ret_val.append(result) 
-                           reasoning = []    
+                    elif var.add_offset.dtype == var.add_offset.dtype and var.scale_factor.dtype != var.dtype:
+                        # Check both attribute types are type float or double
+                        if var.add_offset.dtype in [np.float, np.float16, np.float32, np.float64, np.float128]:
+                            # Check variable types must be byte, short or int
+                            if var.dtype in [np.int8, np.int16, np.int32, np.int64, int()]:
+                                valid = True
+
+                        if valid:
+                            valid_missing_variable = True
+                            # test further with  _FillValue , valid_min , valid_max , valid_range
+                            if hasattr(var, "_FillValue"):
+                                if var._FillValue.dtype !=  var.dtype:
+                                    valid = False
+                                    valid_missing_variable = False
```

Comment:
For each of these, it would be great to start filling out the `reasoning` list with specific messages instead of one message saying "one of these four is wrong".  Since it is a list, it can handle each one as an individual message: 
```python
reasoning.append("_FillValue is set ($dtype) but does not match var dtype ($dtype)")
reasoning.append("valid_min is set ($dtype) but does not match var dtype ($dtype)")
```